sound/soc/sof/intel/hda-loader.c

```diff
@@ -278,10 +282,12 @@ static int hda_cl_cleanup(struct device *dev, struct snd_dma_buffer *dmab,
 			  sd_offset + SOF_HDA_ADSP_REG_SD_BDLPU, 0);
 
 	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, sd_offset, 0);
-	snd_dma_free_pages(dmab);
-	dmab->area = NULL;
-	hstream->bufsize = 0;
-	hstream->format_val = 0;
+	if (!keep_buffer) {
```

Comment:
that's not symmetrical with 

````c
if (!keep_buffer || (keep_buffer && !dmab->area)) {
````

so it's anyone's guess when then memory is freed when the second part of the OR was true.

Reply:
It does not need to be as hda_cl_cleanup cannot check for !dmab->area since if we are here the buffer has been already allocated, it is just a matter of do we want to keep it or not.
I can add a comment here as well.